Sprint 1/step 2

progpspb opened the pull request:
better late than never

progpspb committed:
file structure

progpspb committed:
step-1 final

progpspb committed:
step-2  for review-1

progpspb committed:
burger-constructor useMemo

gennady-bars commented on src/prop-types/ingredient.types.jsx:
```diff
@@ -0,0 +1,17 @@
+import PropTypes from 'prop-types';
+
+const ingredientPropTypes = PropTypes.shape({
```
Отлично, что тип ингредиента вынесен в отдельный файл, чтобы не дублировать его в разных компонентах

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,50 @@
+    const getIngredients = async () => {
+        return fetch(apiUrl)
+            .then(response => {
+                if (!response.ok) {
```
Отлично, что проверяете ответ на `ok`

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,50 @@
+            .then(data => {
+                return data;
+            })
+            .catch(error => console.log(error))
```
Не стоит использовать `catch` тут, так как блок `catch` указан в конце запроса на 32й строчке. Именно там идет отлов всех ошибок, а тут не нужно тормозить их

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,50 @@
+
+    const [ingredients, setIngredients] = useState([]);
+
+    const getIngredients = async () => {
```
Такие функции принято выносить в `api.js`, чтобы не засорять код компонента

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,50 @@
+            try {
+                const result = await getIngredients();
+                setIngredients(result.data);
+            } catch (error) {
```
Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,50 @@
+            }
+        }
+        fetchIngredientsData();
+    }, []);
```
Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,67 @@
+
+const BurgerConstructor = ({ ingredients }) => {
+
+    const [showModal, setModal] = useState(false);
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для управления модальным окном:

```js
import { useState, useCallback } from "react";

// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export const useModal = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  // `useCallback` нужен для того, чтобы зафиксировать ссылку на функцию. Таким образом уменьшится кол-во перерисовок компонента, куда будет передана эта функция
  const openModal = useCallback(() => {
    setIsModalOpen(true);
  }, []);

  const closeModal = useCallback(() => {
    setIsModalOpen(false);
  }, []);

  return {
    isModalOpen,
    openModal,
    closeModal,
  };
};

```

Этот код помещают в отдельный файл `useModal.js` в папке `hooks` и импортируют функцию туда, где нужно создать показатели открытости/закрытости попапа (у каждого попапа они свои)

И теперь нужно просто вызывать хук в том компоненте, где нужно управлять попапом. Всего одна строчка кода будет (под каждый попап своя):

```js
  const { isModalOpen, openModal, closeModal } = useModal();

````

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,67 @@
+
+        </section>
+
+        <Modal isOpen={showModal} setModalOpened={onModalClosed}>
```
Не нужно передавать в пропсы `Modal` показатель открытости `isOpen`, если не делаете **плавное** открытие попапов. Это усложняет логику внутри и усложняет правильное удаление обработчика `Escape`.


Весь попап будет появляться целиком вот так


```jsx

      {showModal && (
        <Modal setModalOpened={onModalClosed}>
          <OrderDetails
            order={{
              number: "034536",
              id: "идентификатор заказа",
              status: "Ваш заказ начали готовить",
              message: "Дождитесь готовности на орбитальной станции",
            }}
          />
        </Modal>
      )}

```

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,94 @@
+            <div className = {styles.tabs}>
+            {ingredientsCollection.map((ingredient, idx) => {
+                return (
+                    <>
```
`<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,94 @@
+            {ingredientsCollection.map((ingredient, idx) => {
+                return (
+                    <>
+                    <Tab value={ingredient.type} active={current === ingredient.type} onClick={setCurrent}>{ingredient.name}</Tab>
```
При отрисовке **массива** компонентов нужно всегда указывать уникальный атрибут `key` в самом **верхнем** уровне верстки.

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,94 @@
+            <div className={styles.tab_scroll + ' pb-10'}>
+                {ingredientsCollection.map((ingredient, idx) => {
+                    return (
+                        <>
+                        <h2 id={ingredient.type} className='text text_type_main-large mt-10 mb-5'>{ingredient.name}</h2>
+                        <div className={styles.items}>
+                        {ingredient.items.map((item, index) => {
+                            return (
+                                <div className={styles.item} key={index} onClick={() => showModalHandler(item)}>
+                                    {index === 0 && <Counter count={1} size="default" /> }
+                                    <img src={item.image_large} alt={item.name} />
+                                    <div className={styles.price}>
+                                        <span>{item.price}</span>
+                                        <CurrencyIcon type="primary" />
+                                    </div>
+                                    <div className='mt-2 mb-6'>{item.name}</div>
+                                </div>
+                            )
+                        })}
+                        </div>
+                        </>
+                    )
```
**Можно лучше**

Лучше это вынести в отдельный компонент `Ingredient`, чтобы код был чище

И тогда `key` тут указать прямо на нем (а то сейчас ошибка)

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,51 @@
+import ModalOverlay from '../modal-overlay/modal-overlay';
+import PropTypes from 'prop-types';
+
+const Modal = ({ title, children, isOpen, setModalOpened }) => {
```
`setModalOpened`

Лучше сразу передавать `onClose`, чтобы в вызов не нужно было бы добавлять `false`. Компонент не должен знать логику закрытия: хоть `false`, хотя другое что-то в будущем

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,51 @@
+
+    const handleOverlayClick = (e) => {
+        e.stopPropagation();
+        setModalOpened(false);
```
`setModalOpened(false)` дублируется 3 раза. Нужно объявить функцию `closeModal` для этого

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,51 @@
+
+    useEffect(() => {
+        const handleCloseEsc = (event) => {
+            if (event.key === 'Escape') setModalOpened(false);
```
`setModalOpened(false)` дублируется 3 раза. Нужно объявить функцию `closeModal` для этого

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,51 @@
+        <div className={styles.modal}>
+            <div className={styles.title}>
+                {title}
+                <span className={styles.close}><CloseIcon type="primary" onClick={() => setModalOpened(false)} /></span>
```
`setModalOpened(false)` дублируется 3 раза. Нужно объявить функцию `closeModal` для этого

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,51 @@
+        isOpen && document.addEventListener('keydown', handleCloseEsc);
+
+        return () => {
+            document.removeEventListener('keydown', handleCloseEsc);
```
Правильно удаляется обработчик

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что ловите возможные ошибки в конце каждого запроса к серверу
* Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз


но есть некоторые недочеты:




* В консоли ошибки https://disk.yandex.ru/i/0hDf3NwfOCqHRw 
* В верхней булке нужно добавлять текст `верх`, а в нижней - `низ`. Посмотрите в фигме, пожалуйста
* Верхняя и нижняя булки должны быть всегда **одинаковыми** по заданию и у них должна быть иконка `Замочек`, так как их нельзя будет удалять (только заменять перетаскиванием)
* **Скролл** нужно сделать между **булками**, чтобы они были всегда видны, а между ними скроллились бы ингредиенты.
* При скролле вниз видна граница оверлея https://disk.yandex.ru/i/zbwisHBmW4T-DQ Модальное окно должно быть всегда ровно посередине экрана. Для этого нужно указать ему и оверлею в `css` позицию `fixed`.
* Не нужно передавать в пропсы `Modal` показатель открытости `isOpen`, если не делаете **плавное** открытие попапов. Это усложняет логику внутри и усложняет правильное удаление обработчика `Escape`.
* `<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку
* При отрисовке **массива** компонентов нужно всегда указывать уникальный атрибут `key` в самом **верхнем** уровне верстки.

progpspb committed:
correct  from review

remove warnings (key, prop)
correct overlay
change burger-constructor
api function

progpspb committed:
correct from review 2

add hook useModal
change Modal usage

progpspb committed:
correct from review

change burger-ingredients
create component ingredient

progpspb committed:
change ingredient component

- useState
+ useModal

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
